src/PowerShell/Microsoft.WinGet.DSC/Microsoft.WinGet.DSC.psm1

```diff
@@ -304,4 +310,175 @@ class WinGetSourcesResource
+        # Installed, need to ensure not installed.
+        if ($installedPackage.Installed -ne $ensureInstalled)
+        {
+            return $false
+        }
+
+        # Different sources.
+        if (-not([string]::IsNullOrWhiteSpace($this.Source)))
+        {
+            if ($this.Source -ne $installedPackage.Source)
+            {
+                return $false
+            }
+        }
+
+        $catalogPackage = Get-WinGetPackage -Id $this.Id -Exact
+
+        # If no version is given, see if the latests is installed.
+        if ([string]::IsNullOrWhiteSpace($this.Version) -and
+            $catalogPackage.IsUpdateAvailable)
+        {
+            return $false
+        }
+
+        # If there is an specific version, compare with the current installed version.
+        if (-not ([string]::IsNullOrWhiteSpace($this.Version)))
+        {
+            $compareResult = $catalogPackage.InstalledVersion.CompareToVersion($this.Version)
+            if ($compareResult -ne 'Equal')
+            {
+                return $false
+            }
+        }
+
+        # For now this is all.
+        return $true
+    }
+
+    # Set.
+    [void] Set()
+    {
+        Assert-WinGetCommand "Install-WinGetPackage"
+        Assert-WinGetCommand "Uninstall-WinGetPackage"
+
+        if (-not $this.Test())
+        {
+            $hashArgs = @{
+                Id = $this.Id
+                Exact = $true
+                Mode = 'Silent'
+            }
+            
+            if ($this.Ensure -eq [Ensure]::Present)
+            {
+                if (-not([string]::IsNullOrWhiteSpace($this.Source)))
+                {
+                    $hashArgs.Add("Source", $this.Source)
+                }
+
+                if ($this.Installed)
+                {
+                    $catalogPackage = Get-WinGetPackage -Id $this.Id -Exact
+
+                    if ([string]::IsNullOrWhiteSpace($this.Version))
+                    {
+                        Update-WinGetPackage @hashArgs
+                    }
+                    else
+                    {
+                        $hashArgs.Add("Version", $this.Version)
+
+                        $compareResult = $catalogPackage.InstalledVersion.CompareToVersion($this.Version)
+                        switch ($compareResult)
+                        {
+                            'Lesser'
+                            {
+                                # Installed package has a lower version. Update.
+                                # TODO: if update fails we should uninstall and install.
+                                Update-WinGetPackage @hashArgs
+                                break
+                            }
+                            {'Greater' -or 'Unknown'}
+                            {
+                                # The installed package has a greated version or unknown. Uninstall and install.
+                                Uninstall-WinGetPackage -Id $this.Id

```

Comment:
We don't have a pipelineable parameter to just take in the package object?  This call is not safe as written (ignores source).